### PR TITLE
orb_exists: change semantics from (is published or subscribed) to (is published)

### PR DESCRIPTION
**Pull request by bkueng**
Existing users of `orb_exists`:
- logger (dynamic subscribe to multi-instances)
- mavlink (orb subscription)
- sdlog2
- preflightcheck (check for home_position)
- wait_for_topic shell command (it's not used)
- orb_group_count() (used in sensors: dynamic sensor addition)
    
All use-cases benefit from the changed semantics: they are really only interested if there is a publisher, not another subscriber.

Additional changes:
- logger: do not subscribe to the first instance if the topic is not advertised (but dynamically check if it's published later on)
- logger: remove unused topics commander_state & rc_channels
- avoid dynamic allocations in orb_exists on POSIX
- mavlink_orb_subscription: reduce orb_exists() check from 10Hz to 3Hz

This has deeper impacts on the overall system, hopefully mostly positive!
The following provides some test results.

### Testing
To get representative results, I tested on a fully assembled quad (pixracer) in armed state with RC and an open QGC connection.

#### Used file descriptors
Number of used file descriptors (mavlink instance is the one that is connected to QGC, without b50b0b7):
```
Before:  mavlink: 46, logger: 54
This PR: mavlink: 30, logger: 33
```

#### Memory (Fragmentation)
The additional file operations (open & close) in `orb_exists` do not lead to memory allocations/frees on Nuttx, but did so on POSIX. I changed this in 3b31a03.

My tests showed a considerable **memory reduction of 4.2 KB**. A bit more than half is because of mavlink, the rest is logger. Since pixracer runs more mavlink instances than other boards, it will be slightly less on other boards.

#### CPU usage
Tests showed this is pretty much unchanged (without 191622b). This is the execution time of `orb_exists` before (with locked scheduling):
```
orb_exists: 23174 events, 879696us elapsed, 37us avg, min 20us max 196us 24.192us rms
```
And this PR:
```
orb_exists: 22182 events, 1383130us elapsed, 62us avg, min 20us max 274us 47.743us rms
```
Obviously slower, but I think it's still ok, since it's not called with high frequency or at critical places.

I suggest reviewing each individual commit.

### CODE REVIEW

File: src/modules/logger/logger.cpp [hunk 1]
```diff
@@ -1070,6 +1107,25 @@ void Logger::run()
 
 #endif /* DBGPRINT */
 
+		} else { // not logging
+
+			// try to subscribe to new topics, even if we don't log, so that:
+			// - we avoid subscribing to many topics at once, when logging starts
+			// - we'll get the data immediately once we start logging (no need to wait for the next subscribe timeout)
+			if (next_subscribe_topic_index != -1) {
+				for (uint8_t instance = 0; instance < ORB_MULTI_MAX_INSTANCES; instance++) {
```
Comment (nicolaerosia) [inline src/modules/logger/logger.cpp]: using uint8_t as index is not an optimization, you should use a counter of processor word size

Comment (bkueng) [reply src/modules/logger/logger.cpp]: I know, I copy-pasted that :/
Fixed in 37902d99de04a329c8

**Comment by dagar** [comment]
This is great! Should we reduce the nuttx fds again?

**Comment by bkueng** [comment]
> This is great! Should we reduce the nuttx fds again?

Thanks! I deliberately did not do that (yet), because:
- Some setups will need more than my numbers above (e.g. VTOL's with camera control, or optical flow setups), however, not as many as before.
- We'll likely use more again in future

If we're short on RAM we can certainly reduce them. One FD needs 16 bytes, so reducing by one implies around 16*20 (tasks) = 320 Bytes more RAM.

**Comment by dagar** [comment]
Memory can be tight in some setups right now, but in most cases it's avoidable. For example it's easy for a user to unintentionally configure a pixracer where they'll run out of memory. Certain combinations of USB connected, SYS_COMPANION, vmount, camera trigger, etc can already leave you without enough memory to successfully calibrate. Not to mention UAVCAN.

It's too bad we can't (easily) configure the nuttx FDs per task. Alternatively we create another uORB interface that doesn't use FDs (for logger and mavlink), but that would need significantly more thought.

Other than mavlink and logger most tasks use < 20 FDs with the exception of commanader, but I already see quite a few ways to bring that down. At what point does it become worth it? Reducing by 20-30 FDs and save 5-10kBs of RAM?

**Comment by LorenzMeier** [comment]
@bkueng Could you look into fixing the CI failures (or rebase) and please ping and tag @PX4TestFlights so they can start testing it? Thanks!

**Comment by bkueng** [comment]
@dagar 
i'd like to see full-blown configs working too on boards like pixracer. But at some point we just hit the limit and I'd suggest to use Pixhawk Pro (it's a really good board too), or fmu-v5 in such cases.

Specifically on Pixracer, we can do one more thing that does not require much effort: `frsky_telemetry` is continuously running (as task), and I think most people don't require it. We could detect if the bus is connected on startup, and if not, exit the task. This would free around 3KB.

How about we reduce the number of FD's by 5? That should definitely be safe.

As for dynamic FD's: I already talked to @davids5 and he told me that Greg had some NuttX testing branch for that, but there were some complications with sockets. So I think it's not going to happen soon. But it's also not that urgent anymore with this PR.
Also the NuttX upgrade requires more RAM, and I think David is looking into reducing that again.
Then together with this PR and @dagar 's navigator optimizations I'd say we're in a good shape again.

CI failed because of flash overflow. I removed the blacksheep telemetry driver from fmu-v2. Circle-CI failure is unrelated and succeeded before.

@PX4TestFlights this is ready for testing. Please test on all boards and look that:
- all the plots for the logs show up correctly, and there are no new error messages
- QGC connection works as expected via USB and telemetry.

**Comment by santiago3dr** [comment]
couple flights on vtol:

https://logs.px4.io/plot_app?log=b8e5a7f0-19ab-4348-84b2-64c179acede2
https://logs.px4.io/plot_app?log=bc5fdf6a-3e42-460a-86af-49536dd2a64c

good flights; no issues; no notable difference from master 747e392
https://logs.px4.io/plot_app?log=7d8e3d14-8aa3-402f-8e74-1fc2288e9eea

**Comment by santiago3dr** [comment]
couple flight on quadcopter running pixracer (v4)

https://logs.px4.io/plot_app?log=cd5708ab-986a-4bed-b650-0b695709c64b
https://logs.px4.io/plot_app?log=f41804b3-106b-429c-80c9-35137e68f312

good flights; no issues

**Pull request merged**